pkg/contexts/ocm/transfer/autohandler.go

```diff
@@ -0,0 +1,111 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package transfer
+
+import (
+	"sync"
+
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/transfer/transferhandler"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/transfer/transferhandler/standard"
+	"github.com/open-component-model/ocm/pkg/errors"
+)
+
+func init() {
+	SetDefaultTransferHandlerFactory(func() transferhandler.TransferHandler { h, _ := standard.New(); return h })
+}
+
+var (
+	lock           sync.Mutex
+	defaultHandler func() transferhandler.TransferHandler
+)
+
+func SetDefaultTransferHandlerFactory(f func() transferhandler.TransferHandler) {
+	lock.Lock()
+	defer lock.Unlock()
+	defaultHandler = f
+}
+
+func NewDefaultTransferHandler() transferhandler.TransferHandler {
+	lock.Lock()
+	defer lock.Unlock()
+	return defaultHandler()
+}
+
+// NewTransferHandler creates a transfer handler for the given set of transfer
+// options. If there is no handler type supporting all the given options
+// an ErrNotSupported error is returned.
+// If no handler can be found according to the origin handlers of given options
+// (an option always belongs to a dedicated option set of a dedicated handler
+// but may be supported by other option sets, also) the options sets of all
+// registered handler types with checked by trying the most significant
+// handlers, first, to find the best matching handler for the given option set.
+func NewTransferHandler(list ...transferhandler.TransferOption) (h transferhandler.TransferHandler, ferr error) {
+	var opts transferhandler.TransferOptions
+
+	created := -1
+outer:
+	for {
```

Comment:
Please try to simplify this three times nested for loop structure by extracting some of the inner components into descriptive functions. Incidentally, that would allow you to skip the labeling of the functions.

Also, the `created` index holder to identify state of the options is hard to follow. Try using a different data structue that would allow for easier reading and following the selection logic of the options.

I feel like the final part of the for loop could be a plain fallback logic on the registered options if none of the given options applied. I'm not sure what makes a second run possible and what would change that warrants only a single retry instead of plainly falling back.

Reply:
The problem is the intrinsic complexity. If the logic of complete sub-loops is split into separate functions, the contract will be very broad, because it has to modify outer state. Additionally the final return is in the inner loops, which makes it difficult to handle this cross function-calls. The result will be much harder to understand than the actual solution.

Reply:
👍